core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategy.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    private static Set<Endpoint> selectRandomly(Set<Endpoint> availableCandidates, int cnt) {
+        final List<Endpoint> endpoints = new ArrayList<>(availableCandidates);
+
+        final Set<Endpoint> selectedCandidates = new HashSet<>();
+        final Random random = ThreadLocalRandom.current();
+
+        for (int i = 0; i < cnt && !endpoints.isEmpty(); i++) {
+            selectedCandidates.add(endpoints.remove(random.nextInt(endpoints.size())));
+        }
+
+        return selectedCandidates;
+    }
+
+    @Override
+    public void updateCandidates(List<Endpoint> candidates) {
+        requireNonNull(candidates, "candidates");
+
+        synchronized (selectedCandidates) {
+            this.candidates = ImmutableSet.copyOf(candidates);
+
+            final Set<Endpoint> removedCandidates = new HashSet<>();
+            for (Endpoint selectedCandidate : selectedCandidates) {
+                if (this.candidates.contains(selectedCandidate)) {
+                    continue;
+                }
+
+                removedCandidates.add(selectedCandidate);
+            }
+
+            updateSelectedCandidates(removedCandidates);
+        }
+    }
+
+    @Override
+    public List<Endpoint> getCandidates() {
+        synchronized (selectedCandidates) {
+            return new ArrayList<>(selectedCandidates);
+        }
+    }
+
+    @Override
+    public boolean updateHealth(Endpoint endpoint, double health) {
+        final double unhealthyScore = 0;
+
+        requireNonNull(endpoint, "endpoint");
+
+        if (!candidates.contains(endpoint)) {
+            unhealthyCandidates.remove(endpoint);
+            return true;
+        }
+
+        if (health > unhealthyScore) {
+            unhealthyCandidates.remove(endpoint);
+            return false;
+        }
+
+        unhealthyCandidates.add(endpoint);
+        synchronized (selectedCandidates) {
+            updateSelectedCandidates(ImmutableSet.of(endpoint));
+            return true;
+        }
+    }
+
+    /*
+    This method must be called with synchronized selectedCandidates
+     */
+    private void updateSelectedCandidates(final Set<Endpoint> removedEndpoints) {
+        final int targetSelectedCandidatesSize = max.calculate(candidates.size());
+
+        removedEndpoints.forEach(selectedCandidates::remove);
+
+        int availableCandidateCount = calculateAvailableCandidateCount(targetSelectedCandidatesSize);
+        if (availableCandidateCount <= 0) {
+            return;
+        }
+
+        final Set<Endpoint> availableCandidates = new HashSet<>(candidates);
+        availableCandidates.removeAll(selectedCandidates);
+        availableCandidates.removeAll(unhealthyCandidates);
+
+        final Set<Endpoint> newSelectedCandidates = selectRandomly(availableCandidates,
+                                                                   availableCandidateCount);
+        selectedCandidates.addAll(newSelectedCandidates);
```

Comment:
We could generalize a `selectRandomly` a little bit more so that:

- we do not need to create a new available candidate `Set`.
- we do not need to create a new selected candidate `Set`.

For example, we could define a method like this:

```java
static void addRandomlySelectedCandidates(
    Set<Endpoint> dest, Set<Endpoint> src, int count,
    Set<Endpoint> exclusions...) {

    final List<Endpoint> endpoints = new ArrayList<>(src.size());
    loop: for (Endpoint e : src) {
        for (Set<Endpoint> exclusion : exclusions) {
            if (exclulsion.contains(e)) {
                continue loop;
            }
        }
        endpoints.add(e);
    }

    final Random random = ThreadLocalRandom.current();
    for (int i = 0; i < count && !endpoints.isEmpty(); i++) {
        dst.add(endpoints.remove(random.nextInt(endpoints.size())));
    }
}
```

Then, we can also simplify the code between L143 and L146.

Reply:
fixed